[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/SmilodonCub/DS4VS/blob/master/Week2/DS4VS_week2_FlowControl.ipynb)

# Loops, Conditionals & Pythonic Alternatives

## Flow Control

<img src="https://www.scientecheasy.com/wp-content/uploads/2019/10/control-flow.png" width="75%" style="margin-left:auto; margin-right:auto">

## Sequential Flow Control 

The Python interpreter executes code line-by-line starting from the top

In [1]:
# an example of some code that will be interpreted sequentially by Python:
import nltk, re, pprint
from nltk.corpus import PlaintextCorpusReader
corpus_root = '/home/bonzilla/Documents/DAACS_BERT/DAACS-Writing/data/WGU-Essays'
wordlists = PlaintextCorpusReader( corpus_root, '.*' )
filenames =  wordlists.fileids()
print( filenames[0] )
file_path = corpus_root + '/' + filenames[0]
file = open( file_path )
Essay0001_raw = file.read( )
len_Essay0001_raw = len( Essay0001_raw )
print( len_Essay0001_raw )

Essay0001.txt
3754


## Problematic!

this folder holds a lot of files:

In [2]:
len( filenames )

1138

What if I want to find the length of each file in the directory?  

Would it be efficient to copy/paste the same code 1138 times?....

# It would be very inefficient, messy, ugly and most certainly unPythonic

## Iterating with `for` loops

basic syntax:

    for an_element in a_sequence:
        execute this code

### `for` loops with lists

In [3]:
print( type( filenames ) )

<class 'list'>


we see that `filenames` is of type `list`, it is an sequence that we can iterate through in a for loop:

In [5]:
for x in filenames:
    file_path = corpus_root + '/' + x
    file = open( file_path )
    raw_essay = file.read( )
    len_raw_essay = len( raw_essay )
    print( len_raw_essay )

3754
2041
1976
2127
2136
2560
2052
2020
2222
3298
2029
2042
1969
2154
2069
3135
2030
2019
2038
2147
1924
1908
2458
2231
2937
3025
2220
2287
1951
2273
3191
2170
2013
2622
2769
2169
1800
2290
2163
2163
2078
1937
2126
2013
2867
2150
2102
2558
2176
2040
2005
1855
1971
1993
1846
2165
3113
2147
2317
2174
1934
2007
2273
1922
4342
2176
2086
2165
1784
2033
3732
2369
2089
2055
1890
2315
2187
2746
2531
2163
4048
1901
2026
2088
2371
1940
2259
2420
2448
1971
1992
2402
1853
2212
1954
2000
1977
2173
2109
2023
1879
2424
2083
2288
2446
2551
2696
3101
2346
2145
1923
2148
2618
2283
3043
2708
1992
2178
3318
2170
2812
2105
2303
3108
2000
1996
2062
1742
1926
2326
1914
2910
2058
1998
2204
2048
2410
1959
1949
2001
1980
3728
3134
2261
2171
3540
2398
2053
2137
1876
1982
3023
2414
2563
2639
1919
2805
2134
2134
2001
1819
2761
2732
2112
2065
2678
2252
1764
3059
1919
3064
2708
1846
5172
2617
2048
2376
1959
2164
3243
1792
1855
2711
2335
1970
1804
5275
3259
3894
2768
2176
2676
1963
2438
1952
1911
1931
2052
2215
2259


What is we want to use these values in the future?  

Let's use this `for` loop to store the essay length values in a new list:

In [8]:
essay_lengths = []
for filename in filenames:
    file_path = corpus_root + '/' + filename
    file = open( file_path )
    raw_essay = file.read( )
    len_raw_essay = len( raw_essay )
    essay_lengths.append(len_raw_essay)
essay_lengths[0:10]

[3754, 2041, 1976, 2127, 2136, 2560, 2052, 2020, 2222, 3298]

### Iterating with a 'range' object

In [11]:
arange = range( 0, 10, 2)
type( arange )

range

In [13]:
num=1
for idx in range( 1, 100, 5 ):
    #print( idx )
    num *= idx
print( num )    

4582077748908899916413273112576


### enumerating `for` loops with lists

In [15]:
#for idx, filename in enumerate( filenames ):
    #print( '{} is offset number {} in filenames'.format( filename, idx ) ) #

In [16]:
# initialize a list to hold the for loop results
name_idx_pairs = []

#loop through using enumerate
for idx, filename in enumerate( filenames ):
    # add the idx & filename values as a tuple to our list
    name_idx_pairs.append(( idx, filename) )

#print the first 10 results
name_idx_pairs[:10]

[(0, 'Essay0001.txt'),
 (1, 'Essay0002.txt'),
 (2, 'Essay0003.txt'),
 (3, 'Essay0004.txt'),
 (4, 'Essay0005.txt'),
 (5, 'Essay0006.txt'),
 (6, 'Essay0007.txt'),
 (7, 'Essay0008.txt'),
 (8, 'Essay0009.txt'),
 (9, 'Essay0010.txt')]

### `for` loops with dictionaries

iteration works well on sequential Python types.  
Can we loop through a dictionary with a for loop?

In [18]:
# let's create a dictionary from the list of tuples we created
name_idx_pairs_dict = {}
for tup in name_idx_pairs:
    name_idx_pairs_dict[ tup[0] ] = tup[1]
    
#name_idx_pairs_dict

In [19]:
# now lets see what happens when we try to iterate through a dictionary
for idx in range(0,10):
    print( name_idx_pairs_dict[ idx ] )

Essay0001.txt
Essay0002.txt
Essay0003.txt
Essay0004.txt
Essay0005.txt
Essay0006.txt
Essay0007.txt
Essay0008.txt
Essay0009.txt
Essay0010.txt


If you put a dictionary directly in a for loop, Python will automatically iterate over the dictionary keys:

In [20]:
for key in name_idx_pairs_dict:
    print( key, 'has the offset: ', name_idx_pairs_dict[ key ])

0 has the offset:  Essay0001.txt
1 has the offset:  Essay0002.txt
2 has the offset:  Essay0003.txt
3 has the offset:  Essay0004.txt
4 has the offset:  Essay0005.txt
5 has the offset:  Essay0006.txt
6 has the offset:  Essay0007.txt
7 has the offset:  Essay0008.txt
8 has the offset:  Essay0009.txt
9 has the offset:  Essay0010.txt
10 has the offset:  Essay0011.txt
11 has the offset:  Essay0012.txt
12 has the offset:  Essay0013.txt
13 has the offset:  Essay0014.txt
14 has the offset:  Essay0015.txt
15 has the offset:  Essay0016.txt
16 has the offset:  Essay0017.txt
17 has the offset:  Essay0018.txt
18 has the offset:  Essay0019.txt
19 has the offset:  Essay0020.txt
20 has the offset:  Essay0021.txt
21 has the offset:  Essay0022.txt
22 has the offset:  Essay0023.txt
23 has the offset:  Essay0024.txt
24 has the offset:  Essay0025.txt
25 has the offset:  Essay0026.txt
26 has the offset:  Essay0027.txt
27 has the offset:  Essay0028.txt
28 has the offset:  Essay0029.txt
29 has the offset:  Essa

In [21]:
# can also iterate through a dictionaries items:
for item in name_idx_pairs_dict.items():
    print( type( item ) )

<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class '

## Iterating with `while` loops

a `while` loop will iterate over code (similar to a `for` loop), but only if a conditional statement is satisfied

basic structure:

    while <conditional statement evaluates True>:
        execute this code

Python Conditionals (Boolean Expressions):

| Relational Operator |          Interpretation         |
|:-------------------:|:-------------------------------:|
|        x == y       |         x is equal to y         |
|        x != y       |       x is not equal to y       |
|        x > y        |       x is greater than y       |
|        x < y        |         x is less than y        |
|        x >= y       | x is greater than or equal to y |
|        x <= y       |   x is less than or equal to y  |

In [22]:
count = 1
while count < 10:
    print( count )
    count += 1

1
2
3
4
5
6
7
8
9


### Controlling `while` loops

If the conditional of a while loop never tests `False`, things can get out of control....  
Your only hope is **CTRL + C**, but in jupyter, you need to stop the kernal. 

    num = 0
    while num < 5:
        print( 'This while loop will run forever!' )

In [23]:
# How do we fix the above while loop?
num = 0
while num < 5:
    print( 'This while loop will run forever!' )
    num += 1

This while loop will run forever!
This while loop will run forever!
This while loop will run forever!
This while loop will run forever!
This while loop will run forever!


### Fine-Tuning loops

Python Logical Operators:

| Logical Operator |          Interpretation         |
|:----------------:|:-------------------------------:|
|      x and y     |      x and y evaluate True      |
|      x or y      |       x or y evaluate True      |
|       not y      |     negation of a statement     |

### Other Control Statements:  

* **`break`** - exit a loop when a statement is met
* **`continue`** - skip the iteration of a loop without exiting
* **`pass`** - allows you to skips some code without exiting the iteration or the loop
* **`try`** - see if some code executes, if not, the user can define methods to handle given the errors thrown by the code.

Before we play with these other control statements, it will be helpful to learn about Conditional Flow Control.....

## Flow Control with Conditionals... using `if`, `elif` & `else`

Conditional `if`/`elif`/`else` statements give your program the ability the reactively change behavior  

Basic structure:  

    if <conditional statement>:
        perform this code
    elif <conditional statement>:
        perform some different code
    else:
        do this instead
        
<br>        
        
* there is no limit to the number of (chained) conditional statements
* there must be an expression following the conditional
    * when no action is needed, `pass` is sufficient

In [27]:
# chained conditional expressions
x = 20
y = 20

if x < y:
    print( 'x is less than y' )
elif x > y: 
    print( 'x is greater than y' )
else:
    print( 'x and y are equal' )

x and y are equal


## Nested Flow Control

Oh my!! ...conditional statements that use `pass` and `break` for added control nested inside a while loop

In [30]:
x = 1
while x <= 20:
    if x > 0 and x <=10: #the exxpression after if is called the conditional
        print( 'x is positive' )
        if x%2==0: #an alternative branch
            print( '{} is even'.format(x) )
            x+=1
        else:
            x+=1
            pass
    elif x > 10:
        print( 'x is positive' )
        if x%2!=0: #an alternative branch
            print( '{} is odd'.format(x) )
            x+=1
        else:
            x+=1
            pass        
    else:
        print( 'x is not positive' )
        break
        

x is positive
x is positive
2 is even
x is positive
x is positive
4 is even
x is positive
x is positive
6 is even
x is positive
x is positive
8 is even
x is positive
x is positive
10 is even
x is positive
11 is odd
x is positive
x is positive
13 is odd
x is positive
x is positive
15 is odd
x is positive
x is positive
17 is odd
x is positive
x is positive
19 is odd
x is positive


## List Comprehensions

Nested conditionals can get quite messy!  
Nested conditionals can take up many lines of code!

Let's remind outselves about the Zen of Python:

In [31]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


### List Comprehensions are a 'Pythonic' alternative  when iterating over lists

general form:

    [<do this!> for an_item in list_of_items ]
    
    [<do this!> for an_item in list_of_items if <condition>]
    
    
Oh joy!:  

* it's pythonic: simple, readable
* it also executes faster than a for loop

In [32]:
hp = """Mr and Mrs Dursley, of number four, Privet Drive, were proud to say that they were perfectly normal, thank you very much. They were the last people you’d expect to be involved in anything strange or mysterious, because they just didn’t hold with such nonsense.
    Mr Dursley was the director of a firm called Grunnings, which made drills. He was a big, beefy man with hardly any neck, although he did have a very large moustache. Mrs Dursley was thin and blonde and had nearly twice the usual amount of neck, which came in very useful as she spent so much of her time craning over garden fences, spying on the neighbours. The Dursleys had a small son called Dudley and in their opinion there was no finer boy anywhere.
    The Dursleys had everything they wanted, but they also had a secret, and their greatest fear was that somebody would discover it. They didn’t think they could bear it if anyone found out about the Potters. Mrs Potter was Mrs Dursley’s sister, but they hadn’t met for several years; in fact, Mrs Dursley pretended she didn’t have a sister, because her sister and her good-for-nothing husband were as unDursleyish as it was possible to be. The Dursleys shuddered to think what the neighbours would say if the Potters arrived in the street. The Dursleys knew that the Potters had a small son, too, but they had never even seen him. This boy was another good reason or keeping the Potters away; they didn’t want Dudley mixing with a child like that."""

print( hp )

Mr and Mrs Dursley, of number four, Privet Drive, were proud to say that they were perfectly normal, thank you very much. They were the last people you’d expect to be involved in anything strange or mysterious, because they just didn’t hold with such nonsense.
    Mr Dursley was the director of a firm called Grunnings, which made drills. He was a big, beefy man with hardly any neck, although he did have a very large moustache. Mrs Dursley was thin and blonde and had nearly twice the usual amount of neck, which came in very useful as she spent so much of her time craning over garden fences, spying on the neighbours. The Dursleys had a small son called Dudley and in their opinion there was no finer boy anywhere.
    The Dursleys had everything they wanted, but they also had a secret, and their greatest fear was that somebody would discover it. They didn’t think they could bear it if anyone found out about the Potters. Mrs Potter was Mrs Dursley’s sister, but they hadn’t met for several y

In [37]:
# make a list of words from the string
hp_words = hp.split(' ')
hp_words

# use a list comprehension to make a list of all the first letters of each word:
hp_1stletters = [ word[0] for word in hp_words if len( word )>0 ]
hp_1stletters

['M',
 'a',
 'M',
 'D',
 'o',
 'n',
 'f',
 'P',
 'D',
 'w',
 'p',
 't',
 's',
 't',
 't',
 'w',
 'p',
 'n',
 't',
 'y',
 'v',
 'm',
 'T',
 'w',
 't',
 'l',
 'p',
 'y',
 'e',
 't',
 'b',
 'i',
 'i',
 'a',
 's',
 'o',
 'm',
 'b',
 't',
 'j',
 'd',
 'h',
 'w',
 's',
 'n',
 'M',
 'D',
 'w',
 't',
 'd',
 'o',
 'a',
 'f',
 'c',
 'G',
 'w',
 'm',
 'd',
 'H',
 'w',
 'a',
 'b',
 'b',
 'm',
 'w',
 'h',
 'a',
 'n',
 'a',
 'h',
 'd',
 'h',
 'a',
 'v',
 'l',
 'm',
 'M',
 'D',
 'w',
 't',
 'a',
 'b',
 'a',
 'h',
 'n',
 't',
 't',
 'u',
 'a',
 'o',
 'n',
 'w',
 'c',
 'i',
 'v',
 'u',
 'a',
 's',
 's',
 's',
 'm',
 'o',
 'h',
 't',
 'c',
 'o',
 'g',
 'f',
 's',
 'o',
 't',
 'n',
 'T',
 'D',
 'h',
 'a',
 's',
 's',
 'c',
 'D',
 'a',
 'i',
 't',
 'o',
 't',
 'w',
 'n',
 'f',
 'b',
 'a',
 'T',
 'D',
 'h',
 'e',
 't',
 'w',
 'b',
 't',
 'a',
 'h',
 'a',
 's',
 'a',
 't',
 'g',
 'f',
 'w',
 't',
 's',
 'w',
 'd',
 'i',
 'T',
 'd',
 't',
 't',
 'c',
 'b',
 'i',
 'i',
 'a',
 'f',
 'o',
 'a',
 't',
 'P',
 'M'

In [39]:
breweries = [ 'singlecut', 'kcbc', 'bridge & tunnel', 'queens', 'finback', 'interboro', 'grimm']

#write a list comprehension to capitalize each item
Breweries = [ brewery.title() for brewery in breweries]
Breweries

['Singlecut',
 'Kcbc',
 'Bridge & Tunnel',
 'Queens',
 'Finback',
 'Interboro',
 'Grimm']

In [41]:
#write a list comprehension to generate a list of strings of odd numbers in a given range
start = 10
stop = 50
odd_range = [ str(num) for num in range( start,stop ) if num%2!=0 and num < 40]
odd_range

['11',
 '13',
 '15',
 '17',
 '19',
 '21',
 '23',
 '25',
 '27',
 '29',
 '31',
 '33',
 '35',
 '37',
 '39']

## Summary

We covered a lot today.  
After working through the basic Python Built-in structures, we took a tour of Flow Control:  

* `for` loops
* `while` loops
* conditional flow cotrol with `it`/`elif`/`else`
* some 'Pythonic' alternatives...
* ...and some other operators  



Flow control is very powerful and adds flexibility to the use of code within a script.  
However, what if we need to reuse chunks of code throughout a script, or across different scripts?   

is it 'Pythonic' to copy/paste loops?......

# NO!!!!

so next week we will look at writing functions and classes as an alternative

## That's all for today!

<img src="https://content.techgig.com/photo/80071467/pros-and-cons-of-python-programming-language-that-every-learner-must-know.jpg?132269" width="100%" style="margin-left:auto; margin-right:auto">